File: pos-saas/prisma/seed-flags.ts
import { PrismaClient } from "../app/generated/prisma/client";
import { PrismaPg } from "@prisma/adapter-pg";
import pg from "pg";
import "dotenv/config";

const pool = new pg.Pool({ connectionString: process.env.DATABASE_URL! });
const adapter = new PrismaPg(pool);
const prisma = new PrismaClient({ adapter });

// ponytail: idempotent upsert by key. Updates name/description/category on
// re-run, but never resets `enabled` (so super-admin toggles survive reseed).
const FLAGS = [
  { key: "dashboard",      name: "Dashboard",                category: "general" },
  { key: "orders",         name: "Orders",                   category: "operations" },
  { key: "customers",      name: "Customers",                category: "operations" },
  { key: "services",       name: "Services & Pricing",       category: "operations" },
  { key: "inventory",      name: "Inventory",                category: "operations" },
  { key: "expenses",       name: "Expenses",                 category: "operations" },
  { key: "deposits",       name: "Customer Deposits",        category: "operations" },
  { key: "pickupRequests", name: "Pickup Requests",          category: "operations" },
  { key: "reports",        name: "Reports",                  category: "operations" },
  { key: "branches",       name: "Outlet Management",        category: "admin" },
  { key: "users",          name: "Staff Management",         category: "admin" },
  { key: "roles",          name: "Role Management",          category: "admin" },
  { key: "billing",        name: "Billing & Subscription",   category: "admin" },
  { key: "website",        name: "Public Website",           category: "growth" },
  { key: "tickets",        name: "Help / Bantuan",           category: "general" },
  { key: "offlineOrderCreate", name: "Offline Order Create", category: "general" },
  { key: "printerSettings", name: "Printer Settings", category: "general" },
  { key: "orderPhotos", name: "Order Proof Photos", category: "operations" },
  { key: "referralProgram", name: "Referral Program", category: "growth" },
  { key: "customersImportExport", name: "Customer CSV Import/Export", category: "operations" },
  { key: "onboardingWizard", name: "Onboarding Wizard", category: "growth" },
  { key: "orderFlowV2", name: "Order Flow V2 (leaner steps)", category: "operations" },
  { key: "staffAttendance", name: "Staff Attendance (clock-in)", category: "operations" },
] as const;

async function main() {
  console.log("🚩 Seeding feature flags...");
  for (const f of FLAGS) {
    await prisma.featureFlag.upsert({
      where: { key: f.key },
      update: { name: f.name, category: f.category },
      create: {
        ...f,
        // ponytail: offlineOrderCreate + staffAttendance default OFF globally —
        // flip per-tenant via super-admin override during dogfood phase.
        enabled: f.key !== "offlineOrderCreate" && f.key !== "staffAttendance",
      },
    });
    console.log(`  ✓ ${f.key}`);
  }
  console.log(`Done. ${FLAGS.length} flags.`);
}

main()
  .catch((e) => {
    console.error(e);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
    await pool.end();
  });

File: hivepos-api/internal/modules/superadmin/seed_flags.go
package superadmin

import (
	"context"
	"log"
)

// SeedFeatureFlags idempotently upserts the platform feature-flag catalog.
// Ports pos-saas/prisma/seed-flags.ts into the Go backend so the super-admin
// flags page + RequireFeatureFlag gate have rows to work with.
//
// ponytail: ON CONFLICT(key) DO UPDATE refreshes name/category but never resets
// `enabled` — super-admin toggles survive reseed (same contract as seed-flags.ts).
// Runs once on boot; best-effort — never fatal.
func (m *Module) SeedFeatureFlags(ctx context.Context) {
	type flag struct {
		key, name, category string
		enabled             bool
	}
	catalog := []flag{
		{"dashboard", "Dashboard", "general", true},
		{"orders", "Orders", "operations", true},
		{"customers", "Customers", "operations", true},
		{"services", "Services & Pricing", "operations", true},
		{"inventory", "Inventory", "operations", true},
		{"expenses", "Expenses", "operations", true},
		{"deposits", "Customer Deposits", "operations", true},
		{"pickupRequests", "Pickup Requests", "operations", true},
		{"reports", "Reports", "operations", true},
		{"branches", "Outlet Management", "admin", true},
		{"users", "Staff Management", "admin", true},
		{"roles", "Role Management", "admin", true},
		{"billing", "Billing & Subscription", "admin", true},
		{"website", "Public Website", "growth", true},
		{"tickets", "Help / Bantuan", "general", true},
		{"offlineOrderCreate", "Offline Order Create", "general", false},
		{"printerSettings", "Printer Settings", "general", true},
		{"orderPhotos", "Order Proof Photos", "operations", true},
		{"referralProgram", "Referral Program", "growth", true},
		{"customersImportExport", "Customer CSV Import/Export", "operations", true},
		{"onboardingWizard", "Onboarding Wizard", "growth", true},
		{"orderFlowV2", "Order Flow V2 (leaner steps)", "operations", true},
		{"staffAttendance", "Staff Attendance (clock-in)", "operations", false},
	}

	const q = `INSERT INTO "FeatureFlag" (id, key, name, category, enabled, "updatedAt")
VALUES (gen_random_uuid()::text, $1, $2, $3, $4, NOW())
ON CONFLICT (key) DO UPDATE SET name = EXCLUDED.name, category = EXCLUDED.category, "updatedAt" = NOW()`

	var n int
	for _, f := range catalog {
		if _, err := m.db.ExecContext(ctx, q, f.key, f.name, f.category, f.enabled); err != nil {
			log.Printf("seed feature-flag %q: %v (continuing)", f.key, err)
			continue
		}
		n++
	}
	log.Printf("✓ Seeded %d/%d feature flags", n, len(catalog))
}
